NDNc/pipeline/pipeline-interests-aimd.cpp

```diff
@@ -0,0 +1,186 @@
+/*
+ * N-DISE: NDN for Data Intensive Science Experiments
+ * Author: Sichen Song <songsichen@cs.ucla.edu>
+ *
+ * MIT License
+ *
+ * Copyright (c) 2021 University of California, Los Angeles
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <iostream>
+
+#include "pipeline-interests-aimd.hpp"
+
+namespace ndnc {
+PipelineInterestsAimd::PipelineInterestsAimd(Face &face, size_t size)
+    : PipelineInterests(face), m_windowSize{size},
+      m_lastDecrease{ndn::time::steady_clock::now()} {}
+
+PipelineInterestsAimd::~PipelineInterestsAimd() {
+    this->end();
+}
+
+void PipelineInterestsAimd::process() {
+    while (this->isValid()) {
+        face->loop();
+        onTimeout();
+
+        if (m_pit->size() >= m_windowSize) {
+            continue;
+        }
+
+        std::vector<PendingInterest> pendingInterests(m_windowSize -
+                                                      m_pit->size());
+        size_t n = m_requestQueue.try_dequeue_bulk(
+            pendingInterests.begin(), m_windowSize - m_pit->size());
+
+        if (n == 0) {
+            continue;
+        }
+
+        std::vector<ndn::Block> interests;
+        interests.reserve(n);
+
+        for (size_t i = 0; i < n; ++i) {
+            auto key = pendingInterests[i].pitEntry;
+
+            m_pit->emplace(std::make_pair(key, pendingInterests[i]));
+            interests.emplace(interests.end(),
+                              std::move(pendingInterests[i].interest));
+
+            m_queue->push(key);
+            m_pit->at(key).markAsExpressed();
```

Comment:
`at()` does another lookup of the item you just inserted, which is inefficient. Use the iterator returned by `emplace()` instead.